twisted/conch/recvline.py

```diff
@@ -323,9 +326,9 @@ def currentHistoryBuffer(self):
 
     def _deliverBuffer(self, buf):
         if buf:
-            for ch in buf[:-1]:
+            for ch in iterbytes(buf[:-1]):
                 self.characterReceived(ch, True)
-            self.characterReceived(buf[-1], False)
+            self.characterReceived(buf[-1:], False)
```

Comment:
This is wrong.  It calls `characterReceived` with a `list` instead of a one-length `bytes`.  Concretely, it breaks Manhole by sometimes sending a list where bytes are expected, producing:

```
2017-02-09 08:47:17-0500 [-] Unhandled Error
        Traceback (most recent call last):
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/python/log.py", line 103, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/python/log.py", line 86, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
            return func(*args,**kw)
        --- <exception caught here> ---
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 597, in _doReadOrWrite
            why = selectable.doRead()
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/internet/process.py", line 280, in doRead
            return fdesc.readFromFD(self.fd, self.dataReceived)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/internet/fdesc.py", line 94, in readFromFD
            callback(output)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/internet/process.py", line 283, in dataReceived
            self.proc.childDataReceived(self.name, data)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/internet/_posixstdio.py", line 77, in childDataReceived
            self.protocol.dataReceived(data)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/insults/insults.py", line 551, in dataReceived
            self._handleControlSequence(b''.join(self.escBuf) + ch)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/insults/insults.py", line 575, in _handleControlSequence
            f(self, self.terminalProtocol, buf[:-1])
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/insults/insults.py", line 602, in B
            handler.keystrokeReceived(proto.DOWN_ARROW, None)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/recvline.py", line 220, in keystrokeReceived
            m()
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/recvline.py", line 355, in handle_DOWN
            self._deliverBuffer(self.historyLines[self.historyPosition])
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/recvline.py", line 331, in _deliverBuffer
            self.characterReceived(buf[-1:], False)
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/manhole.py", line 335, in characterReceived
            source = self.getSource()
          File "/home/exarkun/Environments/txkube/local/lib/python2.7/site-packages/twisted/conch/manhole.py", line 315, in getSource
            ''.join(self.lineBuffer))
        exceptions.TypeError: sequence item 0: expected string, list found

```